library/Opus/Doi/DoiManager.php

```diff
@@ -0,0 +1,667 @@
+            $dois = $doc->getIdentifierDoi();
+            $firstDoi = $dois[0];
+
+            // handelt es sich um eine lokale DOI?
+            if (!$this->isLocalDOI($firstDoi->getValue())) {
+                continue;
+            }
+
+            // hat die lokale DOI den gesuchten Registrierungsstatus
+            $status = $firstDoi->getStatus();
+            if (is_null($statusFilter) ||
+                ($statusFilter == 'unregistered' && is_null($status)) ||
+                ($statusFilter != 'unregistered' && $status == $statusFilter)) {
+                $result[] = $doc;
+            }
+        }
+
+        return $result;
+    }
+
+    /**
+     * Erzeugt auf Basis der konfigurierten DOI-Generator-Klasse einen DOI-Wert für das übergebene Dokument.
+     * Gibt den Wert zurück oder wirft eine Exception, wenn die Generierung nicht möglich ist.
+     *
+     * @param $doc OPUS-Dokument, für das ein DOI-Wert generiert werden soll oder ID eines OPUS-Dokuments
+     *             wird eine ID als String übergeben, so wird versucht das zugehörige Opus_Document aus der Datenbank zu laden
+     *
+     * @throws DoiException
+     */
+    public function generateNewDoi($doc) {
+        // versuche die Generierungsklasse für DOIs zu instanziieren
+        if (!isset($this->config->doi->generatorClass)) {
+            // Fehler: Generierungsklasse für DOIs wurde nicht definiert
+            $message = 'mandatory configuration key doi.generatorClass is missing - check your configuration';
+            $this->defaultLog->err($message);
+            $this->doiLog->err($message);
+            throw new Opus_Doi_DoiException($message);
+        }
+
+        $generatorClassName = $this->config->doi->generatorClass;
+        $classExists = Opus_Util_ClassLoaderHelper::classExists($generatorClassName);
+        
+        if (!$classExists) {
+            // Generierungsklasse für DOI kann nicht gefunden oder geladen werden
+            $message = 'DOI generator class ' . $generatorClassName . ' does not exist or is not instantiable - check configuration';
+            $this->defaultLog->err($message);
+            $this->doiLog->err($message);
+            throw new Opus_Doi_DoiException($message);
+        }
+
+        if (is_string($doc) && is_numeric($doc)) {
+            $docId = $doc;
+            try {
+                $doc = new Opus_Document($docId);
+            }
+            catch (Opus_Model_NotFoundException $e) {
+                $message = 'could not find document ' . $docId . ' in database';
+                $this->defaultLog->err($message);
+                throw new Opus_Doi_DoiException($message);
+            }
+        }
+
+        if (is_null($doc) || !($doc instanceof Opus_Document)) {
+            $message = 'unexpected document class';
+            $this->defaultLog->err($message);
+            throw new Opus_Doi_DoiException($message);
+        }
+
+        try {
+            $generator = new $generatorClassName();
```

Comment:
Ich hätte das in eine Funktion (Factory) ausgelagert, insbesondere da es auch in `Opus_Identifier` verwendet wird und dort mit einem Check, ob die Klasse existiert.

Reply:
ok, guter Hinweis. Baue ich gleich noch ein.